### ISSUE TITLE
pull-kubernetes-integration timing out

### ISSUE THREAD

**Issue by boenn** [kubernetes/kubernetes#103512]
#### Which jobs are flaking:
pull-kubernetes-integration
#### Which test(s) are flaking:

Several packages just started timing out badly after ~7/5

https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&job=integration&test=TestMain

* k8s.io/kubernetes/test/integration/apiserver/apply
* k8s.io/kubernetes/test/integration/auth
* k8s.io/kubernetes/test/integration/client
* k8s.io/kubernetes/test/integration/controlplane
* k8s.io/kubernetes/test/integration/daemonset
* k8s.io/kubernetes/test/integration/disruption

The fact that they all started timing out simultaneously indicates a systemic regression

#### Reason for failure:

Several packages are timing out at 10+ minutes

/kind failing-test
/kind flake
/sig api-machinery

**Comment by liggitt** [comment]
Several packages just started timing out badly after ~7/5

https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&job=pull-kubernetes-integration&xtext=must%20drop%20ALL

* k8s.io/kubernetes/test/integration/apiserver/apply
* k8s.io/kubernetes/test/integration/auth
* k8s.io/kubernetes/test/integration/client
* k8s.io/kubernetes/test/integration/controlplane
* k8s.io/kubernetes/test/integration/daemonset
* k8s.io/kubernetes/test/integration/disruption

The fact that they all started timing out simultaneously indicates a systemic regression

**Comment by liggitt** [comment]
These are the PRs that merged in the past couple days that look like they could potentially impact integration tests:

* #103039 P&F change
* #103437 readyz change
* #99997 ep/epslice refactor (likely unrelated)
* #102998 API server callback in storage critical path
* #103504 P&F change
* #102890 added cmd and integration tests
* #103472 quota integration test deflake (likely unrelated)
* #103484 P&F change

**Comment by liggitt** [comment]
taking a random passing run from before timeouts spiked:

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103493/pull-kubernetes-integration/1412270401798541312

* k8s.io/kubernetes/test/integration/apiserver/apply
  * `<testsuite tests="122" failures="0" time="332.670000" name="k8s.io/kubernetes/test/integration/apiserver/apply">`
* k8s.io/kubernetes/test/integration/auth
  * `<testsuite tests="2213" failures="0" time="362.430000" name="k8s.io/kubernetes/test/integration/auth">`
* k8s.io/kubernetes/test/integration/client
  * `<testsuite tests="68" failures="0" time="419.570000" name="k8s.io/kubernetes/test/integration/client">`
* k8s.io/kubernetes/test/integration/controlplane
  * `<testsuite tests="59" failures="0" time="477.240000" name="k8s.io/kubernetes/test/integration/master">`
* k8s.io/kubernetes/test/integration/daemonset
  * `<testsuite tests="22" failures="0" time="630.090000" name="k8s.io/kubernetes/test/integration/daemonset">`
* k8s.io/kubernetes/test/integration/disruption
  * `<testsuite tests="8" failures="0" time="140.460000" name="k8s.io/kubernetes/test/integration/disruption">`

taking a random failing run from after the timeouts spiked:

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98817/pull-kubernetes-integration/1412827544805183488

* k8s.io/kubernetes/test/integration/apiserver/apply
  * `<testsuite tests="40" failures="0" time="313.420000" name="k8s.io/kubernetes/test/integration/apiserver/apply">`
* k8s.io/kubernetes/test/integration/auth
  * `<testsuite tests="2403" failures="0" time="414.910000" name="k8s.io/kubernetes/test/integration/auth">`
* k8s.io/kubernetes/test/integration/client
  * `<testsuite tests="110" failures="0" time="590.170000" name="k8s.io/kubernetes/test/integration/client">`
* k8s.io/kubernetes/test/integration/controlplane
  * `<testsuite tests="35" failures="0" time="625.220000" name="k8s.io/kubernetes/test/integration/controlplane">`
* k8s.io/kubernetes/test/integration/daemonset
  * `<testsuite tests="22" failures="2" time="1132.490000" name="k8s.io/kubernetes/test/integration/daemonset">`
* k8s.io/kubernetes/test/integration/disruption
  * `<testsuite tests="6" failures="1" time="418.980000" name="k8s.io/kubernetes/test/integration/disruption">`

**Comment by tkashem** [comment]
> E0706 09:25:35.889187   56648 storage_flowcontrol.go:142] "APF bootstrap ensurer is exiting" err="timed out waiting for the condition"

This is a false error message, https://github.com/kubernetes/kubernetes/pull/103550 should address this.

**Comment by liggitt** [comment]
@wojtek-t @tkashem - can you look into the P&F PRs that merged in the last couple days (mentioned in https://github.com/kubernetes/kubernetes/issues/103512#issuecomment-875826569) and see if those increased CPU usage or introduced delays to integration tests?

**Comment by liggitt** [comment]
/assign @wojtek-t @tkashem 

did scale tests indicate any CPU or latency bump in the past couple days?

**Comment by wojtek-t** [comment]
> did scale tests indicate any CPU or latency bump in the past couple days?

Nothing that is visible in the past week.
Though scale-tests are not configured with defaults, which I believe is the case for integration tests.

Looking at the PRs mentioned at https://github.com/kubernetes/kubernetes/issues/103512#issuecomment-875826569 from the P&F area:

* #103039
  * this won't add cpu for sure, but in theory may mean rejecting more requests [but I doubt we have such load in integration tests]
* #102998 
  * We record the call from Count(*) method to store in the cache - but Count(*) is called once per 30s per resource-type and the callback is super cheap, so I really doubt it's this one.
* #103504
  * this is a no-op refactor - shouldn't matter
* #103484
  * this could potentially lead to rejecting requests, but also had one issue is being fixed in https://github.com/kubernetes/kubernetes/pull/103534 - but again I doubt we have such load in integration tests


So the only reason that I can imagine at this point is if we started rejecting more API calls [I don't see any of those imposing a visible cpu usage increase].
Do we have an evidence of increased rejection rate?

**Comment by wojtek-t** [comment]
What is interesting, we're not facing any of these problems in the postsubmit: https://testgrid.k8s.io/sig-release-master-blocking#integration-master

**Comment by wojtek-t** [comment]
BTW - looking more at timing, https://github.com/kubernetes/kubernetes/pull/102998 seem to be the most suspicious one from the timing perspective (from the P&F-related ones). Let me take another look at it.

**Comment by wojtek-t** [comment]
I took another look, I don't see how this could visibly increase usage (and I'm 100% sure it doesn't increase latency).
The only thing it does is it decorates the Count() call to the storage layer (which is called once per minute per type):
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/options/etcd.go#L77
and stores the returned value in cache.
At this point noone is reading from the cache.

So I'm quite sure it has to be something else.

OTOH:
https://github.com/kubernetes/kubernetes/pull/103039 - it seems it was merged much earlier (day+ earlier)
https://github.com/kubernetes/kubernetes/pull/103484 - was merged after the tests started failing
https://github.com/kubernetes/kubernetes/pull/103504 - merged at that time, but it should be no-op refactor

**Comment by tkashem** [comment]
does each "test" in the integration suite spin up a kube-apiserver instance, execute the test and then spin down the apiserver instance? 

I started looking at one failure and I see the following 
```
I0708 13:01:55.468358  123784 httplog.go:104] "HTTP" verb="GET" URI="/healthz" latency="4m6.945361496s" userAgent="Go-http-client/1.1" audit-ID="48c3a4ac-4416-4242-8c34-68a99828300a" srcIP="127.0.0.1:38592" resp=504
```

healthz returning `504` and latency is `4m`. I don't know if this is the predominant error, I am going to continue my investigation.

**Comment by wojtek-t** [comment]
> does each "test" in the integration suite spin up a kube-apiserver instance, execute the test and then spin down the apiserver instance?

Pretty much yes. There is commonly started one etcd (or one etcd per directory? can't remember for sure).

But for other components, tests create them themselves, e.g.:
https://github.com/kubernetes/kubernetes/blob/master/test/integration/client/client_test.go#L59

**Comment by tkashem** [comment]
on another integration failure, https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103539/pull-kubernetes-integration/1413113592659578880 (open PR)

for one test i see
```
testserver.go:332: failed to launch server: failed to wait for /healthz to return ok: timed out waiting for the condition
```

for another failed test
```
testserver.go:235: Waiting for /healthz to be ok...
...
...
F0708 13:04:15.883914  123938 hooks.go:202] PostStartHook "crd-informer-synced" failed: timed out waiting for the condition
```

so the pattern is the kube-apiserver can't initialize in time for the integration

**Comment by tkashem** [comment]
> Testgrid link: link1 and link2
I looked at the link1 test failure - https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103508/pull-kubernetes-integration/1412337676442406912

the kube-apiserver comes up ok, but i do see quite a few 429 (requests rejected by p&f)
```
I0706 09:36:20.865625  126321 httplog.go:104] "HTTP" verb="WATCH" URI="/apis/networking.k8s.io/v1/ingresses?allowWatchBookmarks=true&resourceVersion=49532&timeout=5m47s&timeoutSeconds=347&watch=true" latency="48.454428ms" userAgent="quota.test/v0.0.0 (linux/amd64) kubernetes/$Format" audit-ID="efa58014-d7a4-43cb-811d-7f189c17556e" srcIP="127.0.0.1:37192" resp=429
```

This may be causing the test to time out, But I see this
```
FAIL: TestQuotaLimitService (5.80s)
```
so the test ran for about `6s` and then aborted, but the log doesn't say anything why it aborted.

**Comment by tkashem** [comment]
so i see 159 of these for the test above
```
W0706 09:36:22.034145  126321 apf_controller.go:787] no match found for request &request.RequestInfo{IsResourceRequest:true, Path:"/api/v1/namespaces/quota/resourcequotas/quota/status", Verb:"update", APIPrefix:"api", APIGroup:"", APIVersion:"v1", Namespace:"quota", Resource:"resourcequotas", Subresource:"status", Name:"quota", Parts:[]string{"resourcequotas", "quota", "status"}} and user &user.DefaultInfo{Name:"", UID:"", Groups:[]string(nil), Extra:map[string][]string(nil)}; 
```

the `no match found for request` error is causing p&f to choose `catch-all` - this could be why we are seeing the 429 since catch-all concurrency pool capacity is very limited. 

the `catch-all` is being selected because the caller has no user information `&user.DefaultInfo{Name:"", UID:"", Groups:[]string(nil), Extra:map[string][]string(nil)};`

I think if we increase the concurrency share for `catch-all` that might help, alternatively we could set user information for the caller. let me find out the concurrency limit for `catch-all` in integration setup.

**Comment by liggitt** [comment]
the quota flake is worth looking into, but I'm more concerned about the 5-6 packages that are consistently timing out at 10 minute runs

**Comment by wojtek-t** [comment]
The quota failures were a different problem - and those were fixed: https://github.com/kubernetes/kubernetes/issues/103510

**Comment by tkashem** [comment]
> 5-6 packages that are consistently timing out at 10 minute runs

can you please provide a link to a few failed job i can look at? I looked at link1 and link2 from the issue description.

**Comment by liggitt** [comment]
> can you please provide a link to a few failed job i can look at? I looked at link1 and link2 from the issue description.

https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&job=integration&test=TestMain

**Comment by fromanirh** [comment]
possibly relevant: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/101432/pull-kubernetes-integration/1413156580353904640
```
I0708 15:44:12.288550  122316 etcd.go:73] etcd already running at http://127.0.0.1:2379
FAIL	k8s.io/kubernetes/test/integration/apiserver/apply	665.384s
```

In general this PR (pure kubelet changes) is failing pretty consistently: https://prow.k8s.io/pr-history/?org=kubernetes&repo=kubernetes&pr=101432

**Comment by liggitt** [comment]
yes, k8s.io/kubernetes/test/integration/apiserver/apply is one of the packages consistently timing out. the etcd message is unrelated, that just means the test is using an already-started shared etcd instance.

**Comment by fromanirh** [comment]
got it @liggitt , thanks for clarification and sorry for posting a red herring.

**Comment by tkashem** [comment]
so looking at the logs from `k8s.io/kubernetes/test/integration/apiserver/apply` package that timed out, the following message is immediately before the timeout panic
```
I0708 10:06:29.450093  122428 healthz.go:257] etcd,poststarthook/bootstrap-controller,poststarthook/rbac/bootstrap-roles,poststarthook/scheduling/bootstrap-system-priority-classes,poststarthook/priority-and-fairness-config-producer check failed: healthz

[-]etcd failed: error getting data from etcd: context deadline exceeded

[-]poststarthook/bootstrap-controller failed: not finished

[-]poststarthook/rbac/bootstrap-roles failed: not finished

[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished

[-]poststarthook/priority-and-fairness-config-producer failed: not finished
```

I also see that requests spend more time in the p&f filter (minus the handler execution): 
```
I0708 10:06:13.391113  122428 httplog.go:104] "HTTP" verb="GET" URI="/healthz" latency="10.919842019s" userAgent="Go-http-client/1.1" audit-ID="79048dfa-d22e-4495-91c8-7c1732e6c548" srcIP="127.0.0.1:39726" resp=0 addedInfo="\nauthentication=350.838782ms\nimpersonation=226.198329ms\npriorityandfairness=2.280600827s\nauthorization=244.779767ms"
```

the `/healthz` takes `10s` and it spends `2s` inside the p&f filter `priorityandfairness=2.280600827s`. 

I also see from a failed test the following error
```
I0708 10:12:09.291906  124070 httplog.go:104] "HTTP" verb="PUT" URI="/apis/apps/v1/namespaces/test-rolling-update-deployment/replicasets/test-rolling-update-deployment-58d4bc64c6/status" latency="4m34.258622197s" userAgent="deployment.test/v0.0.0 (linux/amd64) kubernetes/$Format/replicaset-controller" audit-ID="f73655d4-1b7f-4909-b1af-453f63782c84" srcIP="127.0.0.1:59326" resp=504
addedInfo="\nimpersonation=243.636114ms\npriorityandfairness=11.09225205s\nauthorization=1.140282782s\nlogging error output: \"{\\\"metadata\\\":{},\\\"status\\\":\\\"Failure\\\",\\\"message\\\":\\\"Timeout: request did not complete within the allotted timeout\\\",\\\"reason\\\":\\\"Timeout\\\",\\\"details\\\":{},\\\"code\\\":504}\\n\"\n"
```
request timing out with `504`, lateny being reported is  `latency="4m34.258622197s" ` which does not make much sense, it should be below `60s`, also the request spends 11s inside the p&f filetr - `priorityandfairness=11.09225205s`.

I am going to continue my investigation.

**Comment by tkashem** [comment]
```
I0708 10:05:46.649827  122428 storage_factory.go:285] storing events.events.k8s.io in events.k8s.io/v1, reading as events.k8s.io/__internal from storagebackend.Config{Type:"", Prefix:"177e1f16-5a06-43b1-83c5-3fccd9e56537", Transport:storagebackend.TransportConfig{ServerList:[]string{"http://127.0.0.1:2379"}, KeyFile:"", CertFile:"", TrustedCAFile:"", EgressLookup:(egressselector.Lookup)(nil)}, Paging:true, Codec:runtime.Codec(nil), EncodeVersioner:runtime.GroupVersioner(nil), Transformer:value.Transformer(nil), CompactionInterval:300000000000, CountMetricPollPeriod:60000000000, DBMetricPollInterval:30000000000, HealthcheckTimeout:2000000000, LeaseManagerConfig:etcd3.LeaseManagerConfig{ReuseDurationSeconds:60, MaxObjectCount:1000}, ObjectCountTracker:(etcd3.ObjectCountTrackerFunc)(nil)}
```

`etcd3.ObjectCountTrackerFunc` is nil, so i think we can rule out  #102998

**Comment by liggitt** [comment]
looking at when this started according to https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&job=integration&test=TestMain

* k8s.io/kubernetes/test/integration/apiserver/apply
  * 7/6, 17:15 UTC
    * https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/99961/pull-kubernetes-integration/1412459922108452864
    * PR #99961 (not merged until 7/7 1:46 UTC)
  * 7/6, 18:06 UTC
    * https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/99961/pull-kubernetes-integration/1412472575111467008
    * PR #99961 (not merged until 7/7 1:46 UTC)
  * 7/6, 18:27 UTC
    * https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103516/pull-kubernetes-integration/1412477893899980800
    * PR #103516 (showing it existed in master before #99961 merged)

**Comment by wojtek-t** [comment]
> etcd3.ObjectCountTrackerFunc is nil, so i think we can rule out #102998

Yeah - that's a good finding, which I agree rules out that PR.

I think it has to be something that merged immediately before (or right after) 17:00 UTC.
With the above PR ruled out, I don't see any other reasonable candidate.

So I think there are two options:
- either we were lucky for quite some time before and it actually merged earlier [but then why we do observe it so frequently now?]
- or it's actually infra-related change

@liggitt - any other options you can see here?

**Comment by liggitt** [comment]
> the `no match found for request` error is causing p&f to choose `catch-all` - this could be why we are seeing the 429 since catch-all concurrency pool capacity is very limited.

that doesn't look like it's changed since 7/5...

on master:
> go test ./test/integration/apiserver/apply -v | grep "no match found for request" | wc -l
>     197

on f37550a4700:

> go test ./test/integration/apiserver/apply -v | grep "no match found for request" | wc -l
>      197

has the P&F treatment of things that hit the catch-all policy changed since 7/5?

**Comment by liggitt** [comment]
> * or it's actually infra-related change

I looked and there were no modifications to the pull integration test job in the last few days

**Comment by wojtek-t** [comment]
I started looking from a different angle at this now and started comparing the actual `junit.xml` files

I focused on the `integration/apiserver/apply` test.

In healthy runs we have something like this:
```
<testsuite tests="122" failures="0" time="318.020000" name="k8s.io/kubernetes/test/integration/apiserver/apply">
```

In failed runs, I see more like:
```
<testsuite tests="39" failures="0" time="284.339000" name="k8s.io/kubernetes/test/integration/apiserver/apply">
```

I've seen 37, 36, 39 so far.

I compared time to run those tests that were actually finished and those look more-or-less the same in both healthy and failed runs.

So my current hypothesis is that we didn't because of some reason got slower now. The tests that finish seem to be running very similarly (a bit faster or a bit slower).
It's just that at some point something hangs and we stop making any progress. I don't know if it's kube-apiserver that doesn't initialize or something else, but I think it slightly changes the kind of problem we're looking for...

**Comment by wojtek-t** [comment]
> has the P&F treatment of things that hit the catch-all policy changed since 7/5?

Nothing that I'm aware of.

**Comment by liggitt** [comment]
we see the failures on jobs started at 7/6 17:15 UTC and after

first commit after problems appeared in triage:
```
TZ=UTC git show --date=local ea3bcbc

commit ea3bcbc2050b89d11f005a2735c8808e45b8aa21
Merge: 24238425492 1fa673c15c9
Author: Kubernetes Prow Robot <k8s-ci-robot@users.noreply.github.com>
Date:   Tue Jul 6 17:42:19 2021
```

last commit before problems appeared in triage:

```
TZ=UTC git show --date=local 2423842

commit 24238425492227fdbb55c687fd4e94c8b58c1ee3
Merge: dd8ba308771 d0c422fd9c8
Author: Kubernetes Prow Robot <k8s-ci-robot@users.noreply.github.com>
Date:   Tue Jul 6 16:35:02 2021
```

focusing on merges to master in the ~2 hours before 24238425492227fdbb55c687fd4e94c8b58c1ee3

**Comment by wojtek-t** [comment]
The pattern of tests hanging (that I described above in https://github.com/kubernetes/kubernetes/issues/103512#issuecomment-876637353) is actually the same for other failing tests...

**Comment by liggitt** [comment]
Things that merged in the 2-3 hours at/before 2423842:
* #103514
* #103509
* #103504
* #103435
* #103294
* #102520
* #102155
* #103474
* #103462
* #102998

**Comment by rphillips** [comment]
I opened a test revert of [apiserver: add callback to get notified of object count](https://github.com/kubernetes/kubernetes/pull/102998) [here](https://github.com/kubernetes/kubernetes/pull/10359) to see if the new lock impacted performance.

**Comment by wojtek-t** [comment]
As I mentioned above - I'm fairly sure it's not performance-related. It seems like it starts hanging because of some reason (panic? deadlock? ...)

**Comment by wojtek-t** [comment]
> I opened a test revert of 'apiserver: add callback to get notified of object count' over here to see if the new lock impacted performance.

BTW - how you're going to evaluate it? A single success doesn't mean it's that - we're not hard down.

**Comment by liggitt** [comment]
> I opened a test revert of ['apiserver: add callback to get notified of object count'](https://github.com/kubernetes/kubernetes/pull/102998) over [here](https://github.com/kubernetes/kubernetes/pull/103590) to see if the new lock impacted performance.

as noted in https://github.com/kubernetes/kubernetes/pull/102998#discussion_r666449826, that change is not correct anyway... the parameter does not match the documentation or how APF intends to use it

**Comment by rphillips** [comment]
@liggitt https://github.com/kubernetes/kubernetes/pull/102998 doesn't have a delete mechanism either for the map

**Comment by liggitt** [comment]
interestingly, apart from the quota integration flake that was recently introduced and fixed, the post-submit integration job looks healthy: https://testgrid.k8s.io/sig-release-master-blocking#integration-master&width=5

does that run with identical config to the presubmit?

**Comment by liggitt** [comment]
answering my own question:

```diff
35c35
<   job: ci-kubernetes-integration-master
---
>   job: pull-kubernetes-integration
44,45d43
<       - name: SHORT
<         value: --short=false
61c59
<           memory: 20Gi
---
>           memory: 15Gi
64c62
<           memory: 20Gi
---
>           memory: 15Gi
```

**Comment by BenTheElder** [comment]
We should consider increasing the timeout in order to allow PRs to merge given code freeze?

**Comment by rphillips** [comment]
> ```
> I0708 10:05:46.649827  122428 storage_factory.go:285] storing events.events.k8s.io in events.k8s.io/v1, reading as events.k8s.io/__internal from storagebackend.Config{Type:"", Prefix:"177e1f16-5a06-43b1-83c5-3fccd9e56537", Transport:storagebackend.TransportConfig{ServerList:[]string{"http://127.0.0.1:2379"}, KeyFile:"", CertFile:"", TrustedCAFile:"", EgressLookup:(egressselector.Lookup)(nil)}, Paging:true, Codec:runtime.Codec(nil), EncodeVersioner:runtime.GroupVersioner(nil), Transformer:value.Transformer(nil), CompactionInterval:300000000000, CountMetricPollPeriod:60000000000, DBMetricPollInterval:30000000000, HealthcheckTimeout:2000000000, LeaseManagerConfig:etcd3.LeaseManagerConfig{ReuseDurationSeconds:60, MaxObjectCount:1000}, ObjectCountTracker:(etcd3.ObjectCountTrackerFunc)(nil)}
> ```
> 
> `etcd3.ObjectCountTrackerFunc` is nil, so i think we can rule out #102998

This is not accurate. I added two debug statements ([here](https://github.com/kubernetes/kubernetes/pull/102998/files#diff-0ed10b7f1202c452adbdc3c2d3a4c26170bdab72ecb88d320292106c80f1601eR58) and [here](https://github.com/kubernetes/kubernetes/pull/102998/files#diff-0ed10b7f1202c452adbdc3c2d3a4c26170bdab72ecb88d320292106c80f1601eR65)) and the code is called in a hack/local-up-cluster.

**Comment by rphillips** [comment]
@liggitt approved the revert https://github.com/kubernetes/kubernetes/pull/103590 ... the first two tries for pull-kuberentes-integration passed. We may want to see if the revert fixes the issue.

**Comment by liggitt** [comment]
I just trolled through the goroutine dump from a [failure](https://prow.k8s.io/log?container=test&id=1413229880652337152&job=pull-kubernetes-integration) to see if any obvious hung candidates stood out ... nothing lept out at me

```
klog dumping stack as part of fatal log when healthz failed (controlplane_utils.go:248)

goroutine 11892 [running]:
k8s.io/kubernetes/vendor/k8s.io/klog/v2.stacks(0xc000126001, 0xc0029fe3c0, 0x5d, 0x97)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:1026 +0xb9
k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).output(0x7719640, 0xc000000003, 0x0, 0x0, 0xc00bbdc620, 0x1, 0x62fa3eb, 0x15, 0xf8, 0x410300)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:975 +0x1e5
k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).printDepth(0x7719640, 0xc000000003, 0x0, 0x0, 0x0, 0x0, 0x1, 0xc00d119720, 0x1, 0x1)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:735 +0x185
k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).print(...)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:717
k8s.io/kubernetes/vendor/k8s.io/klog/v2.Fatal(...)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:1494
k8s.io/kubernetes/test/integration/framework.startAPIServerOrDie(0xc00afa7500, 0x0, 0x0, 0x0, 0xc007676780, 0x10, 0x10)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/framework/controlplane_utils.go:248 +0xf52
k8s.io/kubernetes/test/integration/framework.RunAnAPIServer(0xc00afa7500, 0xc00afa7500, 0x40fbfb, 0xc00aad74a0)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/framework/controlplane_utils.go:350 +0x45
k8s.io/kubernetes/test/integration/deployment.dcSetup(0xc00ae84300, 0xc00c9e1cb0, 0x45a13a, 0x6271de3, 0x24, 0x6271de3, 0x6271de3, 0x24, 0xc00c9e1d08)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/deployment/util.go:107 +0x69
k8s.io/kubernetes/test/integration/deployment.TestPausedDeployment(0xc00ae84300)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/deployment/deployment_test.go:247 +0x6e
testing.tRunner(0xc00ae84300, 0x4dbc380)
	/usr/local/go/src/testing/testing.go:1193 +0xef
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1238 +0x2b3

klog trying to acquire lock to flush (async):

goroutine 19 [semacquire]:
sync.runtime_SemacquireMutex(0x771965c, 0xc000102d00, 0x1)
	/usr/local/go/src/runtime/sema.go:71 +0x47
sync.(*Mutex).lockSlow(0x7719658)
	/usr/local/go/src/sync/mutex.go:138 +0x105
sync.(*Mutex).Lock(...)
	/usr/local/go/src/sync/mutex.go:81
k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).lockAndFlushAll(0x7719640)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:1176 +0x85
k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).flushDaemon(0x7719640)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:1170 +0x73
created by k8s.io/kubernetes/vendor/k8s.io/klog/v2.init.0
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:420 +0xdf

log healthz check showed klog wasn't blocked:

goroutine 1851 [select]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc002c9d520, 0x538a240, 0xc001c7d3b0, 0x1, 0xc00010e360)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:167 +0x118
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc002c9d520, 0xdf8475800, 0x0, 0x1, 0xc00010e360)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x98
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until(...)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Forever(0xc002c9d520, 0xdf8475800)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:81 +0x4f
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/healthz.(*log).Check.func1
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/healthz/healthz.go:72 +0xf2



test timing out:

goroutine 1 [chan receive, 6 minutes]:
testing.(*T).Run(0xc00ae84300, 0x4c2edf6, 0x14, 0x4dbc380, 0x499801)
	/usr/local/go/src/testing/testing.go:1239 +0x2da
testing.runTests.func1(0xc000388180)
	/usr/local/go/src/testing/testing.go:1511 +0x78
testing.tRunner(0xc000388180, 0xc000c3fcd8)
	/usr/local/go/src/testing/testing.go:1193 +0xef
testing.runTests(0xc0005c6d08, 0x76f4140, 0xe, 0xe, 0xc031f70566a44691, 0x8bbad007a6, 0x7718f80, 0xc00069a660)
	/usr/local/go/src/testing/testing.go:1509 +0x2fe
testing.(*M).Run(0xc0003fe080, 0x0)
	/usr/local/go/src/testing/testing.go:1417 +0x1eb
k8s.io/kubernetes/test/integration/framework.EtcdMain(0xc000c3ff68)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/framework/etcd.go:175 +0x112
k8s.io/kubernetes/test/integration/deployment.TestMain(...)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/deployment/main_test.go:26
main.main()
	_testmain.go:71 +0x18e
```

**Comment by tkashem** [comment]
assuming the tests run serially, I analyzed the tests in `apply` package

```
$ cat apply.log | grep 'PASS:'
--- PASS: TestApplyCRDNoSchema (13.02s)
--- PASS: TestApplyCRDStructuralSchema (12.72s)
--- PASS: TestApplyCRDUnhandledSchema (10.49s)
--- PASS: TestDefaultMissingKeyCRD (11.12s)
--- PASS: TestApplyAlsoCreates (6.60s)
--- PASS: TestNoOpUpdateSameResourceVersion (6.79s)
--- PASS: TestCreateOnApplyFailsWithUID (6.75s)
--- PASS: TestApplyUpdateApplyConflictForced (7.42s)
--- PASS: TestApplyGroupsManySeparateUpdates (6.25s)
--- PASS: TestCreateVeryLargeObject (7.80s)
--- PASS: TestUpdateVeryLargeObject (7.88s)
--- PASS: TestPatchVeryLargeObject (12.86s)
--- PASS: TestApplyManagedFields (7.71s)
--- PASS: TestApplyRemovesEmptyManagedFields (5.88s)
--- PASS: TestApplyRequiresFieldManager (6.12s)
--- PASS: TestApplyRemoveContainerPort (6.08s)
--- PASS: TestApplyFailsWithVersionMismatch (6.84s)
--- PASS: TestApplyConvertsManagedFieldsVersion (6.61s)
--- PASS: TestClearManagedFieldsWithMergePatch (6.28s)
--- PASS: TestClearManagedFieldsWithStrategicMergePatch (6.34s)
--- PASS: TestClearManagedFieldsWithJSONPatch (6.16s)
--- PASS: TestClearManagedFieldsWithUpdate (5.92s)
--- PASS: TestErrorsDontFail (6.71s)
--- PASS: TestErrorsDontFailUpdate (6.16s)
--- PASS: TestErrorsDontFailPatch (7.75s)
--- PASS: TestApplyDoesNotChangeManagedFieldsViaSubresources (6.66s)
--- PASS: TestClearManagedFieldsWithUpdateEmptyList (7.01s)
--- PASS: TestApplyUnsetExclusivelyOwnedFields (6.46s)
--- PASS: TestApplyUnsetSharedFields (7.20s)
--- PASS: TestApplyCanTransferFieldOwnershipToController (5.87s)
--- PASS: TestApplyCanRemoveMapItemsContributedToByControllers (6.85s)
--- PASS: TestDefaultMissingKeys (6.34s)
--- PASS: TestUpgradeClientSideToServerSideApply (6.02s)
--- PASS: TestStopTrackingManagedFieldsOnFeatureDisabled (10.00s)
--- PASS: TestRenamingAppliedFieldManagers (6.09s)
--- PASS: TestRenamingUpdatedFieldManagers (6.23s)
--- PASS: TestDroppingSubresourceField (5.95s)
--- PASS: TestDroppingSubresourceFromSpecField (6.02s)
```

```
$ cat apply.log | grep 'PASS:' | wc -l
38
```

this sums up to about `5m`, and then the 39th test started running and caused the package to timeout, it took about `5m`

```
== RUN   TestSubresourceField

W0708 10:05:44.768336  122428 feature_gate.go:237] Setting GA feature gate ServerSideApply=true. It will be removed in a future release.

...
...

W0708 10:09:55.311561  122428 clientconn.go:1334] [core] grpc: addrConn.createTransport failed to connect to {127.0.0.1:2379 127.0.0.1 <nil> 0 <nil>}: didn't receive server preface in time. Reconnecting...

panic: test timed out after 10m0s
```

**Comment by BenTheElder** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103590/pull-kubernetes-integration/1413245971042668544 just failed on the proposed revert https://github.com/kubernetes/kubernetes/pull/103590

**Comment by liggitt** [comment]
> the `no match found for request` error is causing p&f to choose `catch-all` - this could be why we are seeing the 429 since catch-all concurrency pool capacity is very limited.

fwiw, I checked output from apf_controller.go in integration test stdout logs

on master:
* ~50k lines logged
* ~38k "no match found for request" lines logged

with a [fix](https://github.com/kubernetes/kubernetes/pull/103594) setting a user on integration tests:
* ~13k lines logged
*  ~500 "no match found for request" lines logged
* [still failed](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103594/pull-kubernetes-integration/1413238169389764608) with timeouts on multiple packages

**Comment by liggitt** [comment]
> We should consider increasing the timeout (temporarily) in order to allow PRs to merge given code freeze?

I don't think it's a matter of being just over the edge on passing... the failures I've seen have barely made it through half the tests in a package before just getting stuck

**Comment by tkashem** [comment]
@liggitt  and looking at the log for the test that took `5m` (TestSubresourceField), the following goroutine stack trace is interesting, it is the goroutine that polls to see if the apiserver is running.

```
	var lastHealthContent []byte
	err = wait.PollImmediate(100*time.Millisecond, 30*time.Second, func() (bool, error) {
		result := privilegedClient.Get().AbsPath("/healthz").Do(context.TODO())
		status := 0
		result.StatusCode(&status)
		if status == 200 {
			return true, nil
		}
		lastHealthContent, _ = result.Raw()
		return false, nil
	})
````

```
goroutine 119420 [select, 5 minutes]:

net/http.(*persistConn).roundTrip(0xc00acc19e0, 0xc02c0dd2c0, 0x0, 0x0, 0x0)

	/usr/local/go/src/net/http/transport.go:2610 +0x765

net/http.(*Transport).roundTrip(0x7bb19a0, 0xc027a39300, 0x14830736d69616c, 0x246, 0xc108cfecc0)

	/usr/local/go/src/net/http/transport.go:592 +0xacb

net/http.(*Transport).RoundTrip(0x7bb19a0, 0xc027a39300, 0x4f0e0e3, 0xd, 0xc077a7e200)

	/usr/local/go/src/net/http/roundtrip.go:17 +0x35

k8s.io/kubernetes/vendor/k8s.io/client-go/transport.(*bearerAuthRoundTripper).RoundTrip(0xc018071140, 0xc027a39000, 0xc018071140, 0x0, 0x0)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/transport/round_trippers.go:298 +0x279

net/http.send(0xc027a39000, 0x56dee00, 0xc018071140, 0x0, 0x0, 0x0, 0xc0d3f63628, 0x0, 0x1, 0x0)

	/usr/local/go/src/net/http/client.go:251 +0x454

net/http.(*Client).send(0xc04ecdcc90, 0xc027a39000, 0x0, 0x0, 0x0, 0xc0d3f63628, 0x0, 0x1, 0x43b885)

	/usr/local/go/src/net/http/client.go:175 +0xff

net/http.(*Client).do(0xc04ecdcc90, 0xc027a39000, 0x0, 0x0, 0x0)

	/usr/local/go/src/net/http/client.go:717 +0x45f

net/http.(*Client).Do(...)

	/usr/local/go/src/net/http/client.go:585

k8s.io/kubernetes/vendor/k8s.io/client-go/rest.(*Request).request(0xc026934d00, 0x5740b08, 0xc00005a130, 0xc05b4cf260, 0x0, 0x0)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/rest/request.go:980 +0x42c

k8s.io/kubernetes/vendor/k8s.io/client-go/rest.(*Request).Do(0xc026934d00, 0x5740b08, 0xc00005a130, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/rest/request.go:1038 +0xf1

k8s.io/kubernetes/test/integration/framework.startAPIServerOrDie.func3(0xc017a7de00, 0x0, 0xc017a7de00)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/framework/controlplane_utils.go:236 +0xe5

k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.ConditionFunc.WithContext.func1(0x5740b08, 0xc00005a108, 0xc108cff5a4, 0x1, 0x2)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:217 +0x26

k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext(0x5740b08, 0xc00005a108, 0xc026b9ddb0, 0x0, 0x0, 0x0)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:230 +0x7f

k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.WaitForWithContext(0x5740b08, 0xc00005a108, 0xc0ba516990, 0xc026b9ddb0, 0x0, 0x0)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:657 +0x159

k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.poll(0x5740b08, 0xc00005a108, 0xc0ba516901, 0xc0ba516990, 0xc026b9ddb0, 0x479d1e0, 0xc026b9ddb0)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:591 +0xa5

k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.PollImmediateWithContext(0x5740b08, 0xc00005a108, 0x5f5e100, 0x6fc23ac00, 0xc026b9ddb0, 0x48a2800, 0x573b201)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:523 +0x66

k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.PollImmediate(0x5f5e100, 0x6fc23ac00, 0xc00a154300, 0x0, 0xc104d96b60)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:509 +0x6f

k8s.io/kubernetes/test/integration/framework.startAPIServerOrDie(0xc0f2fa8380, 0x0, 0x0, 0x0, 0xc0003b4e00, 0xc104b25b50, 0xc104b25b20)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/framework/controlplane_utils.go:235 +0xd2a

k8s.io/kubernetes/test/integration/framework.RunAnAPIServer(0xc0f2fa8380, 0xc0f2fa8380, 0xc000142bd8, 0x50d9ea8)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/framework/controlplane_utils.go:350 +0x45

k8s.io/kubernetes/test/integration/apiserver/apply.setup(0x57a1e88, 0xc104a25080, 0x0, 0x0, 0x0, 0xf, 0x7b27001, 0xc104a15080, 0x52ae6f)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/apiserver/apply/apply_test.go:62 +0xd9

k8s.io/kubernetes/test/integration/apiserver/apply.TestSubresourceField(0xc104a25080)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/apiserver/apply/apply_test.go:3344 +0xd7

testing.tRunner(0xc104a25080, 0x50d3ad8)

	/usr/local/go/src/testing/testing.go:1193 +0xef

created by testing.(*T).Run

	/usr/local/go/src/testing/testing.go:1238 +0x2b3
```

**Comment by tkashem** [comment]
so my guess is the apiserver does not start up and the 39th test `TestSubresourceField` times out the package.

Also, i think we should provide a timeout here, so the poll actually returns.
```
result := privilegedClient.Get().AbsPath("/healthz").Do(context.TODO())
```

**Comment by tkashem** [comment]
the above goroutine blocks at `select` here
https://github.com/golang/go/blob/release-branch.go1.16/src/net/http/transport.go#L2610 (i am not sure go 1.16 is the right version)

```
goroutine 119420 [select, 5 minutes]:
````

**Comment by liggitt** [comment]
hoisting more findings from slack

* the passing periodic job is scheduling to the same CI nodes as the failing presubmit job, so the theory that there are "bad nodes" that explain the failures doesn't seem likely
* there are throttled disk IOps on those nodes starting July 6/7 that look likely to be related (not sure if they are a symptom or a cause):
  * ![image](https://user-images.githubusercontent.com/980082/124998997-559ca480-e01b-11eb-90b6-d42bef58d579.png)
  * ![image](https://user-images.githubusercontent.com/980082/124999014-5c2b1c00-e01b-11eb-98c6-c403f15ab869.png)

**Comment by liggitt** [comment]
> Also, i think we should provide a timeout here, so the poll actually returns.
> 
> ```
> result := privilegedClient.Get().AbsPath("/healthz").Do(context.TODO())
> ```

that seems like a good idea, but doesn't explain what changed on 7/6 that made health hang indefinitely

**Comment by tkashem** [comment]
#103437 merged on 07/5, can we rule this out?

**Comment by BenTheElder** [comment]
https://prow.k8s.io/?job=pull-kubernetes-integration success rate is up, but also CI is less loaded this time of day. will be more telling in another 12 hours at least.

**Comment by liggitt** [comment]
pass rate on periodic remains 100% over the last 24 hours: https://testgrid.k8s.io/sig-release-master-blocking#integration-master&width=5

pass rate on presubmit is ~25% (32/125 runs passed): https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-integration&exclude-non-failed-tests=50&width=5

**Comment by tkashem** [comment]
So I took another integration failure and looked into it
```
cat apply.log | grep -E "(PASS:|RUN)"
=== RUN   TestApplyCRDNoSchema
--- PASS: TestApplyCRDNoSchema (13.85s)
=== RUN   TestApplyCRDStructuralSchema
--- PASS: TestApplyCRDStructuralSchema (12.52s)
=== RUN   TestApplyCRDUnhandledSchema
--- PASS: TestApplyCRDUnhandledSchema (9.99s)
=== RUN   TestDefaultMissingKeyCRD
--- PASS: TestDefaultMissingKeyCRD (12.51s)
=== RUN   TestApplyAlsoCreates
--- PASS: TestApplyAlsoCreates (6.35s)
=== RUN   TestNoOpUpdateSameResourceVersion
--- PASS: TestNoOpUpdateSameResourceVersion (7.39s)
=== RUN   TestCreateOnApplyFailsWithUID
--- PASS: TestCreateOnApplyFailsWithUID (7.55s)
=== RUN   TestApplyUpdateApplyConflictForced
--- PASS: TestApplyUpdateApplyConflictForced (6.36s)
=== RUN   TestApplyGroupsManySeparateUpdates
--- PASS: TestApplyGroupsManySeparateUpdates (6.31s)
=== RUN   TestCreateVeryLargeObject
--- PASS: TestCreateVeryLargeObject (6.78s)
=== RUN   TestUpdateVeryLargeObject
--- PASS: TestUpdateVeryLargeObject (7.27s)
=== RUN   TestPatchVeryLargeObject
--- PASS: TestPatchVeryLargeObject (12.44s)
=== RUN   TestApplyManagedFields
--- PASS: TestApplyManagedFields (7.54s)
=== RUN   TestApplyRemovesEmptyManagedFields
--- PASS: TestApplyRemovesEmptyManagedFields (5.94s)
=== RUN   TestApplyRequiresFieldManager
--- PASS: TestApplyRequiresFieldManager (5.80s)
=== RUN   TestApplyRemoveContainerPort
--- PASS: TestApplyRemoveContainerPort (6.40s)
=== RUN   TestApplyFailsWithVersionMismatch
--- PASS: TestApplyFailsWithVersionMismatch (6.53s)
=== RUN   TestApplyConvertsManagedFieldsVersion
--- PASS: TestApplyConvertsManagedFieldsVersion (6.24s)
=== RUN   TestClearManagedFieldsWithMergePatch
--- PASS: TestClearManagedFieldsWithMergePatch (6.55s)
=== RUN   TestClearManagedFieldsWithStrategicMergePatch
--- PASS: TestClearManagedFieldsWithStrategicMergePatch (6.10s)
=== RUN   TestClearManagedFieldsWithJSONPatch
--- PASS: TestClearManagedFieldsWithJSONPatch (5.75s)
=== RUN   TestClearManagedFieldsWithUpdate
--- PASS: TestClearManagedFieldsWithUpdate (6.18s)
=== RUN   TestErrorsDontFail
--- PASS: TestErrorsDontFail (6.99s)
=== RUN   TestErrorsDontFailUpdate
--- PASS: TestErrorsDontFailUpdate (6.47s)
=== RUN   TestErrorsDontFailPatch
--- PASS: TestErrorsDontFailPatch (7.24s)
=== RUN   TestApplyDoesNotChangeManagedFieldsViaSubresources
--- PASS: TestApplyDoesNotChangeManagedFieldsViaSubresources (8.28s)
=== RUN   TestClearManagedFieldsWithUpdateEmptyList
--- PASS: TestClearManagedFieldsWithUpdateEmptyList (6.90s)
=== RUN   TestApplyUnsetExclusivelyOwnedFields
--- PASS: TestApplyUnsetExclusivelyOwnedFields (6.38s)
=== RUN   TestApplyUnsetSharedFields
--- PASS: TestApplyUnsetSharedFields (7.53s)
=== RUN   TestApplyCanTransferFieldOwnershipToController
--- PASS: TestApplyCanTransferFieldOwnershipToController (6.71s)
=== RUN   TestApplyCanRemoveMapItemsContributedToByControllers
--- PASS: TestApplyCanRemoveMapItemsContributedToByControllers (6.36s)
=== RUN   TestDefaultMissingKeys
--- PASS: TestDefaultMissingKeys (6.48s)
=== RUN   TestUpgradeClientSideToServerSideApply
--- PASS: TestUpgradeClientSideToServerSideApply (5.86s)
=== RUN   TestStopTrackingManagedFieldsOnFeatureDisabled
--- PASS: TestStopTrackingManagedFieldsOnFeatureDisabled (9.69s)
=== RUN   TestRenamingAppliedFieldManagers
--- PASS: TestRenamingAppliedFieldManagers (5.89s)
=== RUN   TestRenamingUpdatedFieldManagers
--- PASS: TestRenamingUpdatedFieldManagers (5.84s)
=== RUN   TestDroppingSubresourceField
--- PASS: TestDroppingSubresourceField (6.65s)
=== RUN   TestDroppingSubresourceFromSpecField
--- PASS: TestDroppingSubresourceFromSpecField (6.13s)
=== RUN   TestSubresourceField
```

The first 38 tests run for about `5m` and the 39th test `TestSubresourceField` times out
```
$ cat apply.log | grep 'PASS:' | cut -d " " -f 4 | sed 's/(//g' | sed 's/s)//g' | awk '{s+=$1} END {print s}'
281.75
```

`TestSubresourceField` started running at `07:35:49` and timed out at `07:41:02`
```
=== RUN   TestSubresourceField

W0709 07:35:49.577114  122551 feature_gate.go:237] Setting GA feature gate ServerSideApply=true. It will be removed in a future release.

...
...

W0709 07:41:02.561257  122551 clientconn.go:1334] [core] grpc: addrConn.createTransport failed to connect to {127.0.0.1:2379 127.0.0.1 <nil> 0 <nil>}: didn't receive server preface in time. Reconnecting...

panic: test timed out after 10m0s
```

let me look at the go stack traces to see if i see the same behavior

**Comment by liggitt** [comment]
the test-integration jobs captures etcd debug log in addition to test output... what do we see in the etcd debug log between the last passing test in the package and the package timeout?

**Comment by neolit123** [comment]
Are these timeouts consistantly reproducible locally?

> 127.0.0.1:2379 

That's the etcd endpoint that the kube-apiserver connects to. We did upgrade etcd to 3.5.0 from 3.5.0-rcX recently.

**Comment by liggitt** [comment]
> We did upgrade etcd to 3.5.0 from 3.5.0-rcX recently.

do you have a pointer to when/where that happened? the integration test job uses the etcd in the path, so it is likely the one built into the kubekins-e2e image

**Comment by neolit123** [comment]
the update merged on 7/7 (two days ago):
https://github.com/kubernetes/kubernetes/pull/103142
i do not see any `/retest` commands on the PR.

i was asking about local reproducibility to rule out test cluster node starvation.

**Comment by tkashem** [comment]
I see a go routine associated with a request blocked for `9m`, that should not happen since no test executes for that long, and the hung test times out after about `5m`, so this tells me that a request from an old apiserver instance blocked (it is basically waiting for the watch to be ready, so may be related to etcd). 

it could be a red herring, at the very least the issue is that from one test run to another we have remnants from the previous run. But this test setup has been like this for a long time, it still does not answer why we are seeing the failures now.

looking at the etcd log now. 

```
goroutine 17696 [sync.Cond.Wait, 9 minutes]:

sync.runtime_notifyListWait(0xc0312ff150, 0x0)

	/usr/local/go/src/runtime/sema.go:513 +0xf8

sync.(*Cond).Wait(0xc0312ff140)

	/usr/local/go/src/sync/cond.go:56 +0x99

k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.(*ready).wait(0xc0275637d0)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher/cacher.go:1453 +0x46

k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.(*Cacher).Watch(0xc0312bfb80, 0x5753240, 0xc02ce7c0c0, 0xc020868970, 0xf, 0xc03957ec10, 0x4, 0x0, 0x0, 0x5796978, ...)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher/cacher.go:460 +0x105

k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.(*Cacher).WatchList(0xc0312bfb80, 0x5753240, 0xc02ce7c0c0, 0xc020868970, 0xf, 0xc03957ec10, 0x4, 0x0, 0x0, 0x5796978, ...)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher/cacher.go:542 +0x98

k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry.(*DryRunnableStorage).WatchList(...)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry/dryrun.go:64

k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry.(*Store).WatchPredicate(0xc0312eab40, 0x5753240, 0xc02ce7c0c0, 0x5796978, 0x7c22b00, 0x5796928, 0xc01a0b3098, 0x50e94c0, 0x0, 0x0, ...)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry/store.go:1238 +0x19a

k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry.(*Store).Watch(0xc0312eab40, 0x5753240, 0xc02ce7c0c0, 0xc014ecf9e0, 0x0, 0x0, 0x0, 0x0)

...	

k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithPriorityAndFairness.func1.4.1(0xc028419620, 0x56fea58, 0xc01a0b2fc0, 0x56f52c0, 0xc00dea4cc0, 0x574e1e8, 0xc02d14ab00, 0xc021120000)

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:159 +0xa8

created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithPriorityAndFairness.func1.4

	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:140 +0x25f

```

**Comment by neolit123** [comment]
> pass rate on periodic remains 100% over the last 24 hours:
> pass rate on presubmit is ~25% (32/125 runs passed):

i think we need to check what nodes these job runs landed on, and if the nodes have equal HW specs and available resources during the runs. likely #test-infra would know how to do that.

**Comment by tkashem** [comment]
so these errors are prevalent during the last `5m` which roughly coincides with the window when `TestSubresourceField` started executing.

```
$ cat etcd.log | grep 'http2Server.HandleStreams failed to read frame' | head -n 1
2021-07-09T07:31:11.988Z [transport] transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:2379->127.0.0.1:58162: read: connection reset by peer

$ cat etcd.log | grep 'http2Server.HandleStreams failed to read frame' | tail -n 1
2021-07-09T07:53:25.249Z [transport] transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:2379->127.0.0.1:56624: use of closed network connection

$ cat etcd.log | grep 'http2Server.HandleStreams failed to read frame' | wc -l
1911
```

**Comment by liggitt** [comment]
> i think we need to check what nodes these job runs landed on, and if the nodes have equal HW specs and available resources during the runs. likely #test-infra would know how to do that.
> 
> >

I looked yesterday, and the presubmit and periodic jobs were landing on the same nodes in the same node pool

**Comment by neolit123** [comment]
> transport: http2Server.HandleStreams failed to read frame

there is an open issue about this here:
https://github.com/etcd-io/etcd/issues/9165

but it is qualified as "noisy logs" and not actual problems, i think.

could you share the full log at e.g. pastebin.com?

**Comment by neolit123** [comment]
> I looked yesterday, and the presubmit and periodic jobs were landing on the same nodes in the same node pool

i guess it also depends on how much is running in parallel for the presubmit vs periodic on the nodes.

> so these errors are prevalent during the last 5m which roughly coincides with the window when TestSubresourceField started executing.

i don't see anything other than these grpc warnings in the logs.

we could send an etcd revert PR to an old version and `/test all` it multiple times to see what happens.

supported versions:
https://console.cloud.google.com/gcr/images/k8s-artifacts-prod/US/etcd?gcrImageListsize=30
example PR:
https://github.com/kubernetes/kubernetes/pull/103142

**Comment by tkashem** [comment]
interesting, there is a gap of `5m` nothing in the log.
```
W0709 07:35:51.150450  122551 genericapiserver.go:455] Skipping API node.k8s.io/v1alpha1 because it has no resources.

W0709 07:40:47.683249  122551 clientconn.go:1326] [core] grpc: addrConn.createTransport failed to connect to {127.0.0.1:2379 127.0.0.1 <nil> 0 <nil>}. Err: connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:2379: i/o timeout". Reconnecting...
```

and then a bunch of network error and the test times out.

**Comment by tkashem** [comment]
@neolit123 the logs are here - https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/103465/pull-kubernetes-integration/1413395038607511552/artifacts/

**Comment by liggitt** [comment]
looking at the integration test logs, it looks like integration tests from different packages run in parallel... I'm wondering if we're getting ephemeral port exhaustion or something similar

even though the timelines for flakiness don't line up exactly, the PodSecurity integration tests added a large number of create requests to integration tests in the past week (2554 testcases * 2 object types * 2 server scenarios == ~10k create attempts)

https://github.com/kubernetes/kubernetes/pull/103617 makes those run more sparsely (trims to 458 testcases, so ~2.5k create attempts)

**Comment by neolit123** [comment]
comparing logs between a green periodic run:
https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-integration-master/1413595273233436672
https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-integration-master/1413595273233436672/build-log.txt

and red pull run:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-integration/1413467474686382080
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/103573/pull-kubernetes-integration/1413467474686382080/build-log.txt

i am seeing something odd.

in the red run logs:

```
=== Failed
=== FAIL: test/integration/apiserver/admissionwebhook  (0.00s)
I0709 12:20:19.803851  122539 etcd.go:73] etcd already running at http://127.0.0.1:2379
FAIL	k8s.io/kubernetes/test/integration/apiserver/admissionwebhook	665.147s

=== FAIL: test/integration/apiserver/apply  (0.00s)
I0709 12:20:21.416997  122563 etcd.go:73] etcd already running at http://127.0.0.1:2379
FAIL	k8s.io/kubernetes/test/integration/apiserver/apply	666.295s

=== FAIL: test/integration/auth  (0.00s)
I0709 12:20:28.801504  122638 etcd.go:73] etcd already running at http://127.0.0.1:2379
FAIL	k8s.io/kubernetes/test/integration/auth	665.319s

=== FAIL: test/integration/client  (0.00s)
I0709 12:21:23.598217  122819 etcd.go:73] etcd already running at http://127.0.0.1:2379
FAIL	k8s.io/kubernetes/test/integration/client	605.148s
```

both jobs start etcd from bash:
```
... Starting etcd instance
```

which means that all tests that call EtcdMain -> startEtcd:
https://github.com/kubernetes/kubernetes/blob/3a47ddccdfb09e73758cebb2690ad2a014f67e23/test/integration/framework/etcd.go#L70-L75

such as:
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/test/integration/apiserver/admissionwebhook/main_test.go#L26

would print the string `etcd already running at ...`.
but this string is missing in the logs of the green periodic job.
supposedly, a custom etcd is not started either because the string `starting etcd on` is missing from the logs:
https://github.com/kubernetes/kubernetes/blob/3a47ddccdfb09e73758cebb2690ad2a014f67e23/test/integration/framework/etcd.go#L104

so either something is wrong with the logs of the period job, or the TestMain of certain packages is not being called.

side note, there are more errors in the etcd logs of the green periodic job.

**Comment by liggitt** [comment]
success rate over the last three hours is 100% since https://github.com/kubernetes/kubernetes/pull/103617 merged

**Comment by liggitt** [comment]
https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&job=integration&test=TestMain shows no TestMain failures since that merged as well.

I think more investigation is needed to figure out what tripped the errors, and why the periodic and presubmit jobs showed such dramatically different behavior, but I think the emergency is over for the moment

**Comment by aojea** [comment]
I was only able to took at 2 samples, and it seems this error happens at the same time some of the test fail

> etcd.c82735d6-e12e-11eb-9848-4e35a3fd50f0.root.log.DEBUG.20210710-034543.97065:{"level":"error","ts":"2021-07-10T04:02:46.740Z","caller":"zapgrpc/zapgrpc.go:208","msg":"[transport] transport: Got too many pings from the client, closing the connection.","stacktrace":"go.uber.org/zap/zapgrpc.(*Logger).Errorln\n\t/home/remote/sbatsche/.gvm/pkgsets/go1.16.3/global/pkg/mod/go.uber.org/zap@v1.17.0/zapgrpc/zapgrpc.go:208\ngoogle.golang.org/grpc/internal/grpclog.ErrorDepth\n\t/home/remote/sbatsche/.gvm/pkgsets/go1.16.3/global/pkg/mod/google.golang.org/grpc@v1.38.0/internal/grpclog/grpclog.go:55\ngoogle.golang.org/grpc/grpclog.(*componentData).ErrorDepth\n\t/home/remote/sbatsche/.gvm/pkgsets/go1.16.3/global/pkg/mod/google.golang.org/grpc@v1.38.0/grpclog/component.go:46\ngoogle.golang.org/grpc/grpclog.(*componentData).Errorf\n\t/home/remote/sbatsche/.gvm/pkgsets/go1.16.3/global/pkg/mod/google.golang.org/grpc@v1.38.0/grpclog/component.go:79\ngoogle.golang.org/grpc/internal/transport.(*http2Server).handlePing\n\t/home/remote/sbatsche/.gvm/pkgsets/go1.16.3/global/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_server.go:753\ngoogle.golang.org/grpc/internal/transport.(*http2Server).HandleStreams\n\t/home/remote/sbatsche/.gvm/pkgsets/go1.16.3/global/pkg/mod/google.golang.org/grpc@v1.38.0/internal/transport/http2_server.go:518\ngoogle.golang.org/grpc.(*Server).serveStreams\n\t/home/remote/sbatsche/.gvm/pkgsets/go1.16.3/global/pkg/mod/google.golang.org/grpc@v1.38.0/server.go:918\ngoogle.golang.org/grpc.(*Server).handleRawConn.func1\n\t/home/remote/sbatsche/.gvm/pkgsets/go1.16.3/global/pkg/mod/google.golang.org/grpc@v1.38.0/server.go:868"}

that seems to happen if some clients are not respecting the minimum GRPCKeepAliveMinTime

```
	// GRPCKeepAliveMinTime is the minimum interval that a client should
	// wait before pinging server. When client pings "too fast", server
	// sends goaway and closes the connection (errors: too_many_pings,
	// http2.ErrCodeEnhanceYourCalm). When too slow, nothing happens.
	// Server expects client pings only when there is any active streams
	// (PermitWithoutStream is set false).
	GRPCKeepAliveMinTime time.Duration `json:"grpc-keepalive-min-time"`
```
https://github.com/etcd-io/etcd/blob/beae2e1801663cfc46b73c5d6f4f238713153183/server/embed/config.go#L239-L245

is it possible we have a mismatch between the client grpc versiona and the new etcd 3.5 version?

**Comment by liggitt** [comment]
That looks related to https://github.com/kubernetes/kubernetes/pull/101604

are we enabling an option in the client that requires the server to expect it?

**Comment by pacoxu** [comment]
https://github.com/etcd-io/etcd/issues/8905#issuecomment-346416412
A similar issue: grpc/grpc-go#3171 (comment).

I thought it is only a client configuration. But it seems that configuration in server-side is also needed.
The `--grpc-keepalive-min-time` of etcd should be configured if we set `PermitWithoutStream` to true

https://github.com/etcd-io/etcd/blob/main/server/embed/etcd.go#L716-L722
if `--grpc-keepalive-min-time > 0` , PermitWithoutStream is false on server side.

**Comment by andrewsykim** [comment]
Came here to report that I saw timeout issues on this run https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103645/pull-kubernetes-integration/1414651893304528896 

I noticed in https://github.com/kubernetes/kubernetes/pull/103598 we tried to alleviate the issue by adding a context timeout to /heatlhz check but there's another /healthz check here that I think we missed:

https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/testing/testserver.go#L259-L261

Not fixing the root cause, but do we think it's worthwhile to add a context timeout there too?

**Comment by dims** [comment]
Nothing specific we will do here for 1.22 so just move this along

/milestone v1.23

**Comment by MikeSpreitzer** [comment]
Regarding https://github.com/kubernetes/kubernetes/issues/103512#issuecomment-953314508: that is not a failure, it is normal, and it is much older than 9 days.

**Comment by aojea** [comment]
Just this week in a totally unrelated issue, I've observed go's GC causing long pauses on web servers under stress, we should check if this can be a cause here ...

**Comment by liggitt** [comment]
this started happening again, but without a clear cause. closing this in favor of a fresher investigation issue https://github.com/kubernetes/kubernetes/issues/107857

/close
